### PR TITLE
Preliminary WebXR support

### PR DESCRIPTION
**Pull request by Manishearth**
This implements just enough WebXR to display to 3DOF devices in
immersive mode only.

Couple missing things:

 - [ ] Handling reference spaces (even if just supporting eye-level spaces)
 - [x] Spec links
 - [ ] We enter immersive mode when baseLayer is set, but it seems like
we're supposed to do this when requestSession is called (https://github.com/immersive-web/webxr/issues/453)
 - [ ] VR/XR should block less (https://github.com/servo/servo/issues/22505)
 - [x] More pref-gating
 - [x] `views` is a method instead of an attribute because we don't support FrozenArray

<s>Once I add spec links and pref gating</s> this can be landed as-is for further experimentation.

r? @jdm @MortimerGoro

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22528)

### CODE REVIEW

File: components/script/dom/webidls/VR.webidl [hunk 1]
```diff
@@ -1,11 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
-
-// https://w3c.github.io/webvr/#interface-navigator
-[NoInterfaceObject]
-interface VR {
-  [Pref="dom.webvr.enabled"]
-  Promise<sequence<VRDisplay>> getDisplays();
```
Comment (jdm) [inline components/script/dom/webidls/VR.webidl]: It is not clear to me that eliminating WebVR from our implementation is the right choice right now. It's useful to test against the existing WebVR content.

Comment (Manishearth) [reply components/script/dom/webidls/VR.webidl]: I'm not! `navigator.vr` is old WebVR (it's not in the spec anymore). I still support `navigator.getVRDisplays()` which is the new thing.

Comment (Manishearth) [reply components/script/dom/webidls/VR.webidl]: https://immersive-web.github.io/webvr/spec/1.1/#interface-navigator

Comment (Manishearth) [reply components/script/dom/webidls/VR.webidl]: I just found it convenient to repurpose the VR object as XR since it has the same purpose and I didn't want to duplicate the messaging/syncing logic

**Comment by Manishearth** [comment]
Oh, also, one departure from the spec: we don't support `FrozenArray` and attributes can't be sequences so currently the `views` attribute is a function.

I might try to make it use `any` instead so that it works the same.

**Review by MortimerGoro — CHANGES_REQUESTED** [review]
Great work!

Code looks good on first pass.

File: components/script/dom/webidls/XRFrame.webidl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+// https://immersive-web.github.io/webxr/#xrframe-interface
+
+[SecureContext, Exposed=Window] interface XRFrame {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/webidls/XRFrame.webidl]: Right, I'd mentioned in the PR title that this was missing :smile:

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -266,6 +268,7 @@ impl WebGLRenderingContext {
     pub fn recreate(&self, size: Size2D<u32>) {
         let (sender, receiver) = webgl_channel().unwrap();
         self.webgl_sender.send_resize(size, sender).unwrap();
+        self.size.set(size);
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/webglrenderingcontext.rs]: I believe when that happens it just doesn't work (throwing a GL error). At least, when I was playing with this in Firefox that was the case, there's a limit to the texture size.

**Comment by Manishearth** [comment]
Addressed

File: components/script/dom/xr.rs [hunk 4]
```diff
@@ -53,43 +56,79 @@ impl Drop for XR {
...
+
+        let session = XRSession::new(&self.global(), &displays[0]);
+        promise.resolve_native(&session);
+
+        promise
+    }
+}
+
+impl XR {
+    #[allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/dom/xr.rs]: Why is this required?

File: components/script/dom/xrwebgllayer.rs [hunk 5]
```diff
@@ -3,29 +3,80 @@
...
 
+use std::cell::Cell;
+
 #[dom_struct]
 pub struct XRWebGLLayer {
     xrlayer: XRLayer,
+    antialias: Cell<bool>,
+    depth: Cell<bool>,
+    stencil: Cell<bool>,
+    alpha: Cell<bool>,
```
Comment (jdm) [inline components/script/dom/xrwebgllayer.rs]: Why are these cells?

File: components/script/dom/webglrenderingcontext.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/webglrenderingcontext.rs]: The webgl implementation is allowed to choose a smaller size than what is requested; Servo's implementation does not right now and panics, https://github.com/servo/servo/issues/21718 tracks implementing this correctly.

File: components/script/dom/webidls/XRViewerPose.webidl [hunk 6]
```diff
@@ -7,4 +7,6 @@
 [SecureContext, Exposed=Window] interface XRViewerPose {
   // readonly attribute XRRigidTransform transform;
   // readonly attribute FrozenArray<XRView> views;
+  // workaround until we have FrozenArray
+  sequence<XRView> views();
```
Comment (jdm) [inline components/script/dom/webidls/XRViewerPose.webidl]: Drop a link to https://github.com/servo/servo/issues/10427#issuecomment-449593626.

File: components/script/dom/xrview.rs [hunk 7]
```diff
@@ -44,4 +68,16 @@ impl XRViewMethods for XRView {
     fn Eye(&self) -> XREye {
         self.eye
     }
-}
\ No newline at end of file
+
+    #[allow(unsafe_code)]
+    /// https://immersive-web.github.io/webxr/#dom-xrview-projectionmatrix
+    unsafe fn ProjectionMatrix(&self, _cx: *mut JSContext) -> NonNull<JSObject> {
+        NonNull::new_unchecked(self.proj.get())
```
Comment (jdm) [inline components/script/dom/xrview.rs]: I don't see a good reason to use `new_unchecked` in these methods.

**Review by jdm — APPROVED** [review]
Looks like a useful base implementation!

**Comment by jdm** [comment]
```
error[E0433]: failed to resolve. Use of undeclared type or module `Reflector`
  --> components/script/dom/xrlayer.rs:15:25
   |
15 |             reflector_: Reflector::new(),
   |                         ^^^^^^^^^ Use of undeclared type or module `Reflector`
                                                                                
error[E0412]: cannot find type `Reflector` in this scope
 --> components/script/dom/xrlayer.rs:7:1
  |
7 | #[dom_struct]
  | ^^^^^^^^^^^^^ did you mean `selectors`?
help: possible candidate is found in another module, you can import it into scope
  |
5 | use crate::dom::bindings::reflector::Reflector;
```

File: components/script/dom/xr.rs (hunk 4 above)
Comment (Manishearth) [reply components/script/dom/xr.rs]: anything returning promises trips the lint

https://github.com/servo/servo/issues/22504

Comment (jdm) [reply components/script/dom/xr.rs]: Right, but this doesn't return a promise.

File: components/script/dom/webglrenderingcontext.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/webglrenderingcontext.rs]: Is it enough to just drop a comment linking to that since that's not implemented yet?

File: components/script/dom/xr.rs (hunk 4 above)
Comment (Manishearth) [reply components/script/dom/xr.rs]: oh, I was looking at the wrong impl block

This may be unnecessary, checking

**Comment by Manishearth** [comment]
Updated

@bors-servo try=wpt

File: components/script/dom/webglrenderingcontext.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/webglrenderingcontext.rs]: Sure.

**Comment by jdm** [comment]
```
Diff in /home/travis/build/servo/servo/components/script/dom/xrwebgllayer.rs at line 18:
 use crate::dom::xrviewport::XRViewport;
 use dom_struct::dom_struct;
 
-
 #[dom_struct]
 pub struct XRWebGLLayer {
     xrlayer: XRLayer,
Run `./mach fmt` to fix the formatting
```
